File: Satori-iOS/Satori/Features/Incense/ParticleSystem/SmokeParticleView.swift
import SwiftUI

// MARK: - 烟雾粒子系统
// 轻量级粒子：200–300 个半透明圆点，缓慢上升 + 水平漂移 + 渐隐
// 使用引用类型引擎管理粒子状态，避免在 Canvas 渲染中修改 @State

struct SmokeParticleView: View {
    var burnProgress: CGFloat = 0

    @StateObject private var engine = SmokeParticleEngine()

    var body: some View {
        TimelineView(.animation(minimumInterval: 1.0 / 30.0)) { timeline in
            Canvas { context, size in
                engine.update(canvasSize: size, burnProgress: burnProgress)
                engine.draw(context: context, size: size)
            }
        }
        .allowsHitTesting(false)
    }
}

// MARK: - 粒子引擎（引用类型，安全地在 Canvas 闭包中使用）

final class SmokeParticleEngine: ObservableObject {
    private var particles: [SmokeParticle] = []
    private let maxParticles = 300
    /// 每帧持续生成粒子数（保证烟雾连贯不断）
    private let spawnPerFrame: Int = 2

    func update(canvasSize: CGSize, burnProgress: CGFloat) {
        // 持续生成粒子 —— 每帧固定生成，保证烟雾不间断
        for _ in 0..<spawnPerFrame where particles.count < maxParticles {
            let p = SmokeParticle(
                x: canvasSize.width / 2 + CGFloat.random(in: -4...4),
                y: canvasSize.height,
                vx: CGFloat.random(in: -0.3...0.3),
                vy: CGFloat.random(in: -1.5 ... -0.5),
                radius: CGFloat.random(in: 2...6),
                opacity: Double.random(in: 0.08...0.2),
                life: 1.0,
                decay: Double.random(in: 0.003...0.008)
            )
            particles.append(p)
        }

        // 更新已有粒子
        for i in particles.indices.reversed() {
            particles[i].x += particles[i].vx + sin(particles[i].y * 0.02) * 0.15
            particles[i].y += particles[i].vy
            particles[i].life -= particles[i].decay
            particles[i].radius += 0.03

            if particles[i].life <= 0 {
                particles.remove(at: i)
            }
        }
    }

    func draw(context: GraphicsContext, size: CGSize) {
        for p in particles {
            let alpha = p.opacity * p.life
            guard alpha > 0.005 else { continue }

            let rect = CGRect(
                x: p.x - p.radius,
                y: p.y - p.radius,
                width: p.radius * 2,
                height: p.radius * 2
            )

            var ctx = context
            ctx.opacity = alpha
            ctx.addFilter(.blur(radius: p.radius * 0.5))
            ctx.fill(
                Path(ellipseIn: rect),
                with: .color(.gray.opacity(0.6))
            )
        }
    }
}

// MARK: - 粒子数据

struct SmokeParticle {
    var x: CGFloat
    var y: CGFloat
    var vx: CGFloat
    var vy: CGFloat
    var radius: CGFloat
    var opacity: Double
    var life: Double       // 1.0 → 0.0
    var decay: Double      // 每帧衰减量
}

File: satori_flutter/lib/features/incense/components/smoke_particle_view.dart
import 'dart:math';
import 'package:flutter/material.dart';
import 'package:flutter/scheduler.dart';

/// 烟雾粒子系统
/// 轻量级粒子：最多 300 个半透明圆点，缓慢上升 + 水平漂移 + 渐隐
class SmokeParticleView extends StatefulWidget {
  final double burnProgress;
  final double opacity;

  const SmokeParticleView({
    super.key,
    this.burnProgress = 0,
    this.opacity = 1.0,
  });

  @override
  State<SmokeParticleView> createState() => _SmokeParticleViewState();
}

class _SmokeParticleViewState extends State<SmokeParticleView>
    with SingleTickerProviderStateMixin {
  late final Ticker _ticker;
  final _engine = _SmokeEngine();

  @override
  void initState() {
    super.initState();
    _ticker = createTicker((_) {
      _engine.burnProgress = widget.burnProgress;
      setState(() {});
    });
    _ticker.start();
  }

  @override
  void dispose() {
    _ticker.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Opacity(
      opacity: widget.opacity,
      child: CustomPaint(
        painter: _SmokePainter(_engine),
        size: Size.infinite,
      ),
    );
  }
}

class _SmokeEngine {
  final List<_SmokeParticle> particles = [];
  static const _maxParticles = 300;
  static const _spawnPerFrame = 2;
  double burnProgress = 0;
  final _rng = Random();

  void update(Size size) {
    // 生成
    for (var i = 0; i < _spawnPerFrame && particles.length < _maxParticles; i++) {
      particles.add(_SmokeParticle(
        x: size.width / 2 + _rng.nextDouble() * 8 - 4,
        y: size.height,
        vx: _rng.nextDouble() * 0.6 - 0.3,
        vy: -(_rng.nextDouble() * 1.0 + 0.5),
        radius: _rng.nextDouble() * 4 + 2,
        opacity: _rng.nextDouble() * 0.12 + 0.08,
        life: 1.0,
        decay: _rng.nextDouble() * 0.005 + 0.003,
      ));
    }

    // 更新
    for (var i = particles.length - 1; i >= 0; i--) {
      final p = particles[i];
      p.x += p.vx + sin(p.y * 0.02) * 0.15;
      p.y += p.vy;
      p.life -= p.decay;
      p.radius += 0.03;
      if (p.life <= 0) {
        particles.removeAt(i);
      }
    }
  }
}

class _SmokeParticle {
  double x, y, vx, vy, radius, opacity, life, decay;
  _SmokeParticle({
    required this.x,
    required this.y,
    required this.vx,
    required this.vy,
    required this.radius,
    required this.opacity,
    required this.life,
    required this.decay,
  });
}

class _SmokePainter extends CustomPainter {
  final _SmokeEngine engine;
  _SmokePainter(this.engine);

  @override
  void paint(Canvas canvas, Size size) {
    engine.update(size);

    final paint = Paint()..maskFilter = null;

    for (final p in engine.particles) {
      final alpha = p.opacity * p.life;
      if (alpha < 0.005) continue;

      paint
        ..color = Colors.grey.withValues(alpha: (0.6 * alpha))
        ..maskFilter = MaskFilter.blur(BlurStyle.normal, p.radius * 0.5);

      canvas.drawOval(
        Rect.fromCenter(
          center: Offset(p.x, p.y),
          width: p.radius * 2,
          height: p.radius * 2,
        ),
        paint,
      );
    }
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) => true;
}
